### ISSUE TITLE
Provide tip-of-trunk C++ compilers

### ISSUE THREAD

**Issue by gnzlbg** [travis-ci/travis-ci#3579]
It would be nice to have tip-of-trunk compilers (clang and gcc) in both linux and mac. Right now, testing C++ libraries with new compilers is a world of pain under Travis (configuring/installing llvm/gcc libc++/libstdc++ is a world of pain in macos and linux). Then there is the issue that the Ubuntu LTS packages just can't keep up, there is a new C++ standard every 3 years, the release cycle of LTS is 2 years, and it takes at least an extra year for Travis to update the Ubuntu LTS version. This is just not enough. Finally, there is a discussion in the Boost mailing list (incubator of C++ libraries that are then proposed for standardization) to require future libraries proposed for peer review to be tested under travis. Most of the new libraries require C++14 (this means clang 3.7 and gcc 5.0).

All this problems would be solved if the nightlies (2 compilers: gcc, clang, 2 standard libraries: libstdc++, libc++) where available in the virtual machines. That would make Travis much more attractive for testing C++ libraries overnight.

**Comment by michaelbprice** [comment]
This is rather important.  I am also developing libraries (and language extensions) that will be proposed for standardization, and at least the most recent stable release is critical.  It could be risky to go off of the tip-of-trunk, but the benefits are very high.

**Comment by grooverdan** [comment]
providing two compilers:
<pre>
compiler:
  - gcc
  - clang
</pre>

for later versions
<pre>
addons:
  apt:
    sources:
      - ubuntu-toolchain-r-test
      - llvm-toolchain-precise-3.6
    packages:
      - gcc-5
      - g++-5
      - clang-3.6

before_install:
  - sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-5 50
  - sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-5 50
  - sudo update-alternatives --install /usr/bin/gcov gcov /usr/bin/gcov-5 50
  - if [ ${CC} == "clang" ]; then
    export CC=/usr/bin/clang-3.6 CXX=/usr/bin/clang++-3.6;
    fi
</pre>

The clang-3.7 seems to have been replaced by llvm-toolchain-snapshot recently  (http://llvm.org/apt/precise/pool/main/l/)  which will require a sudo add-apt-repository -y .... line in before_install.

libstdc++, libc++ could be handled with:

<pre>
env:
  matrix:
    LIBC=libc
    LIBC=libstdc++

before_install
  - if [ ${LIBC} == "libstdc++" ]; then
       sudo apt-get install ....  ;
    fi
</pre>

**Comment by grooverdan** [comment]
using repository llvm-toolchain-precise and package clang-3.7 will provide 3.7.0-svn239191

**Comment by gnzlbg** [comment]
Do you happen to have a small github project that showcases this to compile a hello world? 

I think the best solution we have is [what Hana does](https://github.com/ldionne/hana/blob/master/.travis.yml), but as you see there is no easy way to install e.g. libc++. 

AFAIK there is also no easy way to install gcc 5.1, see [this issue](https://github.com/travis-ci/travis-ci/issues/3736). 

If you manage to easily install gcc 5.1 or find an easy way (e.g. using apt-get) for installing libstdc++ or libc++ please let us know.

**Comment by grooverdan** [comment]
haven't tried libstdc++/libc++

gcc-5 is gcc-5.1 is exactly like i described.

https://github.com/openquery/mariadb-server/blob/travis-ci-5.5/.travis.yml

https://travis-ci.org/openquery/mariadb-server/jobs/65860320

gcc (Ubuntu 5.1.0-0ubuntu11~12.04.2) 5.1.0

**Comment by gnzlbg** [comment]
@grooverdan awesome Daniel, will give this a try later!

**Comment by ldionne** [comment]
It's working for me right now. See [this](https://travis-ci.org/ldionne/hana/jobs/68701715#L321): GCC 5 is being installed and used properly. The build fails because of a compilation error, but that's another story.

**Comment by gnzlbg** [comment]
This issue has not been solved.

**Comment by gnzlbg** [comment]
This issue has not been solved.

**Comment by BanzaiMan** [comment]
@gnzlbg Could you elaborate on what you are asking for, that is not provided by the solution @grooverdan showed?

**Comment by gnzlbg** [comment]
I'd like to just use `clang-trunk` `gcc-trunk` as the compilation toolchain without having to add llvm's and gcc's ubuntu's package repositories manually and fetching their trunk toolchain versions. Testing against the current development version of the C and C++ compilers ought to be that easy. 

@grooverdan solution is just a work around for this, which breaks pretty much every 6 months when the compiler version of the development branches change, the package repositories change, etc.

**Comment by BanzaiMan** [comment]
> FWIW I think that testing against the current development version of the C and C++ compilers ought to be as easy as just writing `clang-trunk` or `gcc-trunk` in the `.travis.yml`.

I agree. However, it does not appear to me that there is enough tooling to make this happen easily. If you can help with this, or you know someone who can, we appreciate your input.

**Comment by Mizux** [comment]
Concerning "clang" (with CMake based build) is not so straight forward.
* Clang use the wrong gcc headers instead of clang-5.0.0 ones -> see cf #9550.
* Clang by default use `"-stdlib=libstdc++"` which [seems not working](https://travis-ci.org/Mizux/abseil-cpp/jobs/446210290#L555) so you have to add `-stdlib=libc++` on command line (when using `-std=c++11`)

Except that everything "just work" out the box IMHO...
